DFE-Digital/sts-plan-technology-for-your-school
# Feat: New qa visualiser

## Overview

Addresses ticket [#238548](https://dfe-ssp.visualstudio.com/s190-schools-technology-services/_workitems/edit/238548)

Updates the qa-visualiser to use a new api endpoint introduced in the cms controller to get sections. Also makes some simplifications now that the sections come back in a nested format

## Changes

### Web app
- New CMS endpoint for retrieving Sections
  - Methods to map answer nextQuestions within a section to a question containing only system details. 
    - This is to avoid an infinite reference loop that happens despite the JSON options 
  - Endpoint uses same API key auth as the webhook

### QA visualiser
- Simplifies the qa-visualiser to use the new nested structure for sections
- Adds type hinting to all files
- Uses pydantic to serialize and validate data coming back from the cms endpoint
- Previously if an answer ended the topic, its arrow to the "Check Answers" node would be labelled with "No more questions". This has been changed to the answer text to be consistent with all the other arrows and more clear.
- Has just one entry point `main.py` and does the image generation in one go
- Uses uv for project dependencies, linting and formatting
- Unit tests in pytest

### Github Actions
- Github action takes a branch rather than an environment, with the option to include or exclude preview contents. So to get staging content, you'd run from branch main including preview content, and for production without preview content
  - (you can only run it from this branch until merged though, so only with dev content)
- Github action for getting db connection string has been changed to generic secret retrieval so it can be used for other secrets such as api key value
- Github action for qa-visualiser updated to:
  - put the api key value in the environment vars
  - Remove database connection steps
  - Run the visualisation in one go
- New github action for qa-visualiser unit tests and linting running only when this project changes

## How to review the PR
1. Run plantech locally
2. Setup your .env file with the api--authentication--value from the keyvault
3. Follow the Readme instructions for the qa-visualiser and make sure it works
4. Run the github pipeline and check it outputs here too
5. Check that ticking and unticking the include preview content works as expected

## Additional Notes

I considered calling Contentful directly with the python contentful library rather than going via a plan tech api endpoint - I think it would be a nice thing to do, perhaps following on. Unfortunately if you use an inclusion depth of higher than 1 the contentful client just crashes. I also tried retrieving questions, answers and sections at depth 1 to connect manually but it frequently timed out. So a cli export that the app parses might be the way to go

## Checklist
